### PR TITLE
Handle arithmetic overflow in select_addr

### PR DESCRIPTION
**Pull request by gretay-js**
Selection performs constant folding on memory addresses for amd64. Large constants can be created with flambda2 that cause overflow. This PR uses nativeint instead of int. 
It's possible that better code can be generated if we recognize earlier that constants overflow and fold some of them rather than bailing out on the entire expression. I'll add a cmm-based test.

### CODE REVIEW

**Comment by lthls** [comment]
Just to make sure I understand this PR: this does not prevent overflows, it merely hopes that overflows on the target will be replicated exactly by `Nativeint`. Wouldn't it be better to use `Int64` instead, as we know it's going to match the size of the actual operations even in the case of cross-compilation (assuming we ever support it) ?
My preferred fix would be to consider any overflow in displacement a red flag, and use `Misc.no_overflow_*` to check for overflow, with either a warning or just the default case when we do detect an overflow.
But I guess your comment about folding part of the expression refers to when we end up with displacements that are outside the range of allowed immediates ? That looks like an interesting improvement, but not very much related to overflows.

**Comment by gretay-js** [comment]
> My preferred fix would be to consider any overflow in displacement a red flag, and use `Misc.no_overflow_*` to check for overflow, with either a warning or just the default case when we do detect an overflow. But I guess your comment about folding part of the expression refers to when we end up with displacements that are outside the range of allowed immediates ? That looks like an interesting improvement, but not very much related to overflows.

Use of Misc.no_overflow_  was my first version... Nativeint is slightly simpler but if correctness is more obvious without it, I'm happy to change. 

 What I meant  in the optimization is to use is_immediate directly before every fold, which obviates the need for overflow checking on amd64 (I'm ignoring crosscompilation for the moment)

**Comment by gretay-js** [comment]
> Wouldn't it be better to use `Int64` instead, as we know it's going to match the size of the actual operations even in the case of cross-compilation (assuming we ever support it) ?

For cross-compilation, it would be Targetint not Int64, right?

**Comment by lthls** [comment]
> > Wouldn't it be better to use `Int64` instead, as we know it's going to match the size of the actual operations even in the case of cross-compilation (assuming we ever support it) ?
> 
> For cross-compilation, it would be Targetint not Int64, right?

Both would work, since we're in the `amd64` backend. `Targetint` does feel more natural, indeed.

**Comment by gretay-js** [comment]
I modified the code to use Misc.no_overflow_*.

**Comment by lthls** [comment]
I'm not sure why you chose to use a Cmm test, I can reproduce the problem quite reliably with the following ml file:
```ocaml
let[@inline never][@local never] f n =
  let n = Int64.of_int n in
  let open Int64 in
  to_int (add n (of_int Int.min_int))

let _ = Printf.printf "%d\n%!" (f 1)
```
I think adding a simple `(* TEST *)` header + the reference output should be enough to integrate it in the testsuite.

**Comment by gretay-js** [comment]
> I'm not sure why you chose to use a Cmm test, I can reproduce the problem quite reliably with the following ml file:

I replaced the cmm test with your example, it's so much better, thank you! 

Just for the recorded, the output before this PR was `0x1`. Correct output is `0x4000000000000001`.

File: backend/amd64/selection.ml [hunk 1]
```diff
@@ -31,54 +31,60 @@ type addressing_expr =
...
-           ((Asymbol _ | Aadd (_, _) | Ascaledadd (_, _, _)), _))
+           ((Alinear _ | Asymbol _ | Aadd (_, _) | Ascaledadd (_, _, _)), _))
         | (((Asymbol _ | Aadd (_, _) | Ascaledadd (_, _, _)), _),
            ((Asymbol _ | Alinear _ | Aadd (_, _) | Ascaledadd (_, _, _)), _))
           ->
               (Aadd(arg1, arg2), 0)
       end
   | arg ->
-      (Alinear arg, 0)
+    (Alinear arg, 0)
```
Comment (xclerc) [inline backend/amd64/selection.ml:87]: (Could be `| _ -> default`; not sure which
version is easier to read.)

File: backend/amd64/selection.ml [hunk 2]
```diff
@@ -31,54 +31,60 @@ type addressing_expr =
...
+          when Misc.no_overflow_add n1 n2 ->
               (Aadd(e1, e2), n1 + n2)
-        | ((Alinear e1, n1), (Ascale(e2, scale), n2)) ->
+        | ((Alinear e1, n1), (Ascale(e2, scale), n2))
+          when Misc.no_overflow_add n1 n2 ->
               (Ascaledadd(e1, e2, scale), n1 + n2)
-        | ((Ascale(e1, scale), n1), (Alinear e2, n2)) ->
+        | ((Ascale(e1, scale), n1), (Alinear e2, n2))
+          when Misc.no_overflow_add n1 n2 ->
               (Ascaledadd(e2, e1, scale), n1 + n2)
```
Comment (xclerc) [inline backend/amd64/selection.ml:74]: (Possibly merge these cases, as it was done for
`Cadd`_k_ and `Cmuli` left/right above.)

File: ocaml/testsuite/tests/asmcomp/select_addr.ml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+(* TEST
+
```
Comment (xclerc) [inline ocaml/testsuite/tests/asmcomp/select_addr.ml:2]:
```suggestion
```

File: backend/amd64/selection.ml [hunk 4]
```diff
@@ -223,7 +229,8 @@ method select_addressing _chunk exp =
   (* PR#4625: displacement must be a signed 32-bit immediate *)
   if not (is_immediate d)
   then (Iindexed 0, exp)
-  else match a with
+  else
+    match a with
```
Comment (xclerc) [inline backend/amd64/selection.ml:233]: (Could be reverted.)

**Review by xclerc — APPROVED** [review]

File: backend/amd64/selection.ml (hunk 4 above)
Comment (gretay-js) [reply backend/amd64/selection.ml]: Fixed, thanks!

File: backend/amd64/selection.ml (hunk 1 above)
Comment (gretay-js) [reply backend/amd64/selection.ml]: Fixed.

File: backend/amd64/selection.ml (hunk 2 above)
Comment (gretay-js) [reply backend/amd64/selection.ml]: Done (note the swap in argument names e1 e2 here).

File: backend/amd64/selection.ml [hunk 5]
```diff
@@ -31,54 +31,57 @@ type addressing_expr =
...
       end
   | Cop((Caddi | Caddv | Cadda), [arg1; arg2], _) ->
       begin match (select_addr arg1, select_addr arg2) with
-          ((Alinear e1, n1), (Alinear e2, n2)) ->
+          ((Alinear e1, n1), (Alinear e2, n2))
+          when Misc.no_overflow_add n1 n2 ->
               (Aadd(e1, e2), n1 + n2)
-        | ((Alinear e1, n1), (Ascale(e2, scale), n2)) ->
+        | (((Alinear e1, n1), (Ascale(e2, scale), n2)) |
+           ((Ascale(e2, scale), n1), (Alinear e1, n2)))
```
Comment (xclerc) [inline backend/amd64/selection.ml:70]:
```suggestion
           ((Ascale(e2, scale), n2), (Alinear e1, n1)))
```

**Review by mshinwell — APPROVED** [review]

**Pull request merged**